src/lib/collab/channel.js

```diff
@@ -1,76 +1,180 @@
-      this.emit('telepointer', payload)
-    })
-    this.eventEmitter.emit('connected', {
-      doc,
-      version
-    })
+  resetBackoff() {
+    this.backoff = 0
+    this.failures = 0
   }
 
-  debounce(getState) {
-    if (this.debounced) {
-      clearTimeout(this.debounced)
+  /**
+   * Increase the backoff
+   */
+  increaseBackoff() {
+    this.failures = this.failures + 1
+    if (this.failures % failuresBeforeCatchup == 0) {
+      this.emit('needcatchup', { failures: this.failures })
     }
 
-    this.debounced = window.setTimeout(() => {
-      this.sendSteps(getState(), getState)
-    }, 250)
+    this.backoff = Math.max(
+      Math.min(this.backoff * 2, maximumBackoff),
+      minimumBackoff
+    )
   }
 
   /**
-   * Send steps to service
+   * Program something after backoff
+   * @returns {Promise}
    */
-  async sendSteps(state, getState, localSteps) {
-    const { docId } = this.config
-
-    if (this.isSending) {
-      this.debounce(getState)
-      return
+  afterBackoff() {
+    const fn = resolve => {
+      if (this.backoff < minimumBackoff) {
+        resolve()
+      } else {
+        window.setTimeout(resolve, this.backoff)
+      }
     }
+    return new Promise(fn)
+  }
 
-    const version = getVersion(state)
+  /**
+   * Initialize the steps queue
+   */
+  initializeStepsQueue() {
+    this.queuedStep = undefined
+  }
 
-    // Don't send any steps before we're ready.
-    if (typeof version === undefined) {
-      return
-    }
+  /**
+   * Enqueue steps
+   * @param {Function} getState - function to get a current proseMirror state
+   * @param {Object} state - proseMirror state
+   * @param {Object[]} localSteps - local steps to send
+   */
+  enqueueSteps({ getState, state, localSteps }) {
+    this.queuedStep = { getState, state, localSteps }
+  }
 
-    const { steps } = localSteps || sendableSteps(state) || { steps: [] } // sendableSteps can return null..
+  /**
+   * Dequeue steps
+   * @returns {getState, state, localSteps}
+   */
+  dequeueSteps() {
+    const queued = this.queuedStep
+    const getState = queued.getState
+    const state = queued.state || getState()
+    const localSteps = queued.localSteps ||
+      sendableSteps(state) || { steps: [] }
+    this.queuedStep = undefined
+
+    return { getState, state, localSteps }
+  }
 
-    if (steps.length === 0) {
-      return
+  /**
+   * Test if there is a queued step
+   * @returns {boolean}
+   */
+  hasQueuedSteps() {
+    return !!this.queuedStep
+  }
+
+  /**
+   * Rebase steps in queue
+   */
+  rebaseStepsInQueue() {
+    if (this.queuedStep) {
+      this.queuedStep = { getState: this.queuedStep.getState }
     }
+  }
+
+  /**
+   * Program a push for new steps to the service
+   * @param {Function} getState - function to get a current proseMirror state
+   * @param {Object} state - proseMirror state
+   * @param {Object[]} localSteps - local steps to send
+   */
+  async sendSteps(getState, state, localSteps) {
```

Comment:
Are we using this `localSteps` attribute somewhere?

Reply:
In `dequeueSteps`, requested by `processQueue`

Reply:
Documentation added